src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.InProcOrRemoteHostAnalyzerRunner.cs

```diff
@@ -12,145 +12,219 @@
-            {
-                _asyncOperationListener = operationListener;
-                _analyzerInfoCache = analyzerInfoCache;
-            }
+        public Task<DiagnosticAnalysisResultMap<DiagnosticAnalyzer, DiagnosticAnalysisResult>> AnalyzeDocumentAsync(
+            DocumentAnalysisScope documentAnalysisScope,
+            CompilationWithAnalyzers compilationWithAnalyzers,
+            bool logPerformanceInfo,
+            bool getTelemetryInfo,
+            CancellationToken cancellationToken)
+            => AnalyzeAsync(documentAnalysisScope, documentAnalysisScope.TextDocument.Project, compilationWithAnalyzers,
+                forceExecuteAllAnalyzers: false, logPerformanceInfo, getTelemetryInfo, cancellationToken);
+
+        public Task<DiagnosticAnalysisResultMap<DiagnosticAnalyzer, DiagnosticAnalysisResult>> AnalyzeProjectAsync(
+            Project project,
+            CompilationWithAnalyzers compilationWithAnalyzers,
+            bool forceExecuteAllAnalyzers,
+            bool logPerformanceInfo,
+            bool getTelemetryInfo,
+            CancellationToken cancellationToken)
+            => AnalyzeAsync(documentAnalysisScope: null, project, compilationWithAnalyzers,
+                forceExecuteAllAnalyzers, logPerformanceInfo, getTelemetryInfo, cancellationToken);
+
+        private async Task<DiagnosticAnalysisResultMap<DiagnosticAnalyzer, DiagnosticAnalysisResult>> AnalyzeAsync(
+            DocumentAnalysisScope? documentAnalysisScope,
+            Project project,
+            CompilationWithAnalyzers compilationWithAnalyzers,
+            bool forceExecuteAllAnalyzers,
+            bool logPerformanceInfo,
+            bool getTelemetryInfo,
+            CancellationToken cancellationToken)
+        {
+            var result = await AnalyzeCoreAsync().ConfigureAwait(false);
+            Debug.Assert(getTelemetryInfo || result.TelemetryInfo.IsEmpty);
+            return result;
 
-            public async Task<DiagnosticAnalysisResultMap<DiagnosticAnalyzer, DiagnosticAnalysisResult>> AnalyzeAsync(CompilationWithAnalyzers compilation, Project project, bool forcedAnalysis, CancellationToken cancellationToken)
+            async Task<DiagnosticAnalysisResultMap<DiagnosticAnalyzer, DiagnosticAnalysisResult>> AnalyzeCoreAsync()
             {
-                Contract.ThrowIfFalse(compilation.Analyzers.Length != 0);
+                Contract.ThrowIfFalse(!compilationWithAnalyzers.Analyzers.IsEmpty);
 
                 var workspace = project.Solution.Workspace;
-                var service = workspace.Services.GetService<IRemoteHostClientProvider>();
-                if (service == null)
+                if (workspace.Services.GetService<IRemoteHostClientProvider>() is { } service &&
+                    await service.TryGetRemoteHostClientAsync(cancellationToken).ConfigureAwait(false) is { } remoteHostClient)
```

Comment:
Use `RemoteHostClient.TryGetClientAsync` instead.

Reply:
Thanks, that simplified the checks.